server/eclipse-project/src/main/java/us/freeandfair/corla/model/Round.java

```diff
@@ -253,45 +258,76 @@ public Integer startAuditedPrefixLength() {
   }
   
   /**
-   * @return the number of discrepancies.
+   * Ensures that a counter exists in the specified map for the specified key.
+   * 
+   * @param the_map The map.
+   * @param the_key The key.
+   */
+  private void ensureCounterExists(final Map<AuditReason, Integer> the_map,
+                                   final AuditReason the_key) {
+    if (!the_map.containsKey(the_key)) {
+      the_map.put(the_key, 0);
+    }
+  }
+  
+  /**
+   * @return the numbers of discrepancies found in the audit so far, 
+   * categorized by contest audit reason.
    */
-  public Integer discrepancies() {
-    return my_discrepancies;
+  public Map<AuditReason, Integer> discrepancies() {
+    return Collections.unmodifiableMap(my_discrepancies);
   }
   
   /**
-   * Adds a discrepancy.
+   * Adds a discrepancy for the specified audit reason. This adds it both to the 
+   * total and to the current audit round, if one is ongoing.
+   * 
+   * @param the_reason The reason.
    */
-  public void addDiscrepancy() {
-    my_discrepancies = my_discrepancies + 1;
+  public void addDiscrepancy(final AuditReason the_reason) {
+    ensureCounterExists(my_discrepancies, the_reason);
+    my_discrepancies.put(the_reason, my_discrepancies.get(the_reason) + 1); 
   }
   
   /**
-   * Removes a discrepancy.
+   * Removes a discrepancy for the specified audit reason. This removes it 
+   * both from the total and from the current audit round, if one is ongoing.
+   * 
+   * @param the_reason The reason.
    */
-  public void removeDiscrepancy() {
-    my_discrepancies = my_discrepancies - 1;
+  public void removeDiscrepancy(final AuditReason the_reason) {
+    ensureCounterExists(my_discrepancies, the_reason);
+    my_discrepancies.put(the_reason, my_discrepancies.get(the_reason) - 1);
```

Comment:
Won't this result in negative discrepancies if `the_reason` was not in `my_discrepancies`, since `ensureCounterExists()` sets the value of `my_discrepancies.get(the_reason)` to 0 in that case?

Reply:
Yes. Yes it will. Rounds have always been able to have negative discrepancies because if you _resubmit_ a ballot that had previously caused a discrepancy, during a different round than you submitted it, the discrepancy counter still goes down. This is a complete nonissue during regular use because we never resubmit the same ballot twice, and could potentially be confusing if we did... but is correct behavior.

Reply:
I see. I hadn't realized this because we don't observe it on the client.